firmware/ci.c

```diff
@@ -404,23 +413,156 @@ static void video_mode_list(void)
 
 static void video_mode_set(int mode)
 {
-	char mode_descriptors[PROCESSOR_MODE_COUNT*PROCESSOR_MODE_DESCLEN];
+	char mode_descriptor[PROCESSOR_MODE_DESCLEN];
 	if(mode < PROCESSOR_MODE_COUNT) {
-		processor_list_modes(mode_descriptors);
-		wprintf("Setting video mode to %s\r\n", &mode_descriptors[mode*PROCESSOR_MODE_DESCLEN]);
+		processor_describe_mode(mode_descriptor, mode);
+		wprintf("Setting video mode to %s\r\n", mode_descriptor);
 		config_set(CONFIG_KEY_RESOLUTION, mode);
 		processor_start(mode);
 	}
 }
 
+#define NEXT_TOKEN_OR_RETURN(s, t)				\
+	if(!(t = get_token(&s))) {				\
+		wprintf("Parse failed - invalid mode.\r\n");	\
+		return;						\
+	}
+
+static void video_mode_custom(char* str)
```

Comment:
Can you add a github issue for you to refactor this so it can be tested outside the firmware (maybe in a similar way to the edid test?)